File: php/ExtractFunction.php
<?php

declare(strict_types=1);

namespace RefactoringPatterns;

/**
 * Extract Function Refactoring Pattern
 *
 * This refactoring pattern involves taking a fragment of code that can be grouped together
 * and turning it into a function whose name explains the purpose of the function.
 *
 * Benefits:
 * - Improves code readability by giving meaningful names to code fragments
 * - Reduces code duplication by extracting reusable logic
 * - Makes code easier to understand and maintain
 * - Facilitates testing by isolating specific behaviors
 *
 * When to use:
 * - When you have a code fragment that can be grouped together
 * - When you need to use the same code fragment in multiple places
 * - When a function is too long and does multiple things
 * - When you need to add comments to explain what code is doing
 *
 * @see https://refactoring.com/catalog/extractFunction.html
 */
class ExtractFunction
{
    /**
     * BEFORE: Long method with mixed responsibilities
     * Problems:
     * - Hard to understand what the method does at a glance
     * - Multiple levels of abstraction mixed together
     * - Difficult to test individual pieces of logic
     * - Code duplication potential
     */
    public function printOwingBefore(array $invoice): void
    {
        $outstanding = 0;

        echo "********************\n";
        echo "*** Customer Owes ***\n";
        echo "********************\n";

        // calculate outstanding
        foreach ($invoice['orders'] as $order) {
            $outstanding += $order['amount'];
        }

        // record due date
        $today = new \DateTimeImmutable();
        $invoice['dueDate'] = $today->modify('+30 days');

        // print details
        echo "name: {$invoice['customer']}\n";
        echo "amount: {$outstanding}\n";
        echo "due: {$invoice['dueDate']->format('Y-m-d')}\n";
    }

    /**
     * AFTER: Refactored method using Extract Function pattern
     * Benefits:
     * - Clear, self-documenting code with meaningful function names
     * - Each function has a single responsibility
     * - Easier to test each piece independently
     * - Better code organization and maintainability
     */
    public function printOwingAfter(array $invoice): void
    {
        $this->printBanner();
        $outstanding = $this->calculateOutstanding($invoice);
        $this->recordDueDate($invoice);
        $this->printDetails($invoice, $outstanding);
    }

    /**
     * Extracted function: Print banner
     * Single responsibility: Display the header banner
     */
    private function printBanner(): void
    {
        echo "********************\n";
        echo "*** Customer Owes ***\n";
        echo "********************\n";
    }

    /**
     * Extracted function: Calculate outstanding amount
     * Single responsibility: Sum up all order amounts
     *
     * @param array $invoice The invoice data
     * @return float The total outstanding amount
     */
    private function calculateOutstanding(array $invoice): float
    {
        $outstanding = 0;
        foreach ($invoice['orders'] as $order) {
            $outstanding += $order['amount'];
        }
        return $outstanding;
    }

    /**
     * Extracted function: Record due date
     * Single responsibility: Set the due date for the invoice
     *
     * @param array &$invoice The invoice data (passed by reference)
     */
    private function recordDueDate(array &$invoice): void
    {
        $today = new \DateTimeImmutable();
        $invoice['dueDate'] = $today->modify('+30 days');
    }

    /**
     * Extracted function: Print invoice details
     * Single responsibility: Display customer and payment information
     *
     * @param array $invoice The invoice data
     * @param float $outstanding The outstanding amount
     */
    private function printDetails(array $invoice, float $outstanding): void
    {
        echo "name: {$invoice['customer']}\n";
        echo "amount: {$outstanding}\n";
        echo "due: {$invoice['dueDate']->format('Y-m-d')}\n";
    }

    /**
     * Example usage demonstrating the refactoring
     */
    public function demonstratePattern(): void
    {
        $invoice = [
            'customer' => 'John Doe',
            'orders' => [
                ['amount' => 100.00],
                ['amount' => 250.50],
                ['amount' => 75.25],
            ],
        ];

        echo "=== BEFORE Refactoring ===\n";
        $this->printOwingBefore($invoice);

        echo "\n=== AFTER Refactoring ===\n";
        // Reset invoice for clean comparison
        $invoice = [
            'customer' => 'John Doe',
            'orders' => [
                ['amount' => 100.00],
                ['amount' => 250.50],
                ['amount' => 75.25],
            ],
        ];
        $this->printOwingAfter($invoice);

        echo "\n=== Key Benefits ===\n";
        echo "1. Each function has a clear, single purpose\n";
        echo "2. Code is self-documenting through function names\n";
        echo "3. Individual functions can be reused and tested independently\n";
        echo "4. Easier to modify specific behaviors without affecting others\n";
    }
}

File: rust/src/extract_function.rs
/// # Extract Function Refactoring Pattern
///
/// Take a fragment of code that can be grouped together and turn it into
/// a function whose name explains the purpose.
///
/// ## When to use
/// - When you have a code fragment that can be grouped together
/// - When a function is too long and does multiple things
/// - When you need to add comments to explain what code is doing
///
/// Reference: <https://refactoring.com/catalog/extractFunction.html>

use std::fmt;

struct Order {
    amount: f64,
}

struct Invoice {
    customer: String,
    orders: Vec<Order>,
    due_date: Option<String>,
}

impl fmt::Display for Invoice {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "name: {}\namount: {:.2}\ndue: {}",
            self.customer,
            self.orders.iter().map(|o| o.amount).sum::<f64>(),
            self.due_date.as_deref().unwrap_or("N/A")
        )
    }
}

// === BEFORE: Mixed responsibilities in one function ===

fn print_owing_before(invoice: &mut Invoice) {
    let mut outstanding = 0.0;

    println!("********************");
    println!("*** Customer Owes ***");
    println!("********************");

    // calculate outstanding
    for order in &invoice.orders {
        outstanding += order.amount;
    }

    // record due date
    invoice.due_date = Some("2024-02-14".to_string()); // simplified for demo

    // print details
    println!("name: {}", invoice.customer);
    println!("amount: {outstanding:.2}");
    println!("due: {}", invoice.due_date.as_deref().unwrap());
}

// === AFTER: Each piece extracted into its own function ===

fn print_banner() {
    println!("********************");
    println!("*** Customer Owes ***");
    println!("********************");
}

fn calculate_outstanding(invoice: &Invoice) -> f64 {
    invoice.orders.iter().map(|o| o.amount).sum()
}

fn record_due_date(invoice: &mut Invoice) {
    invoice.due_date = Some("2024-02-14".to_string());
}

fn print_details(invoice: &Invoice, outstanding: f64) {
    println!("name: {}", invoice.customer);
    println!("amount: {outstanding:.2}");
    println!("due: {}", invoice.due_date.as_deref().unwrap());
}

fn print_owing_after(invoice: &mut Invoice) {
    print_banner();
    let outstanding = calculate_outstanding(invoice);
    record_due_date(invoice);
    print_details(invoice, outstanding);
}

fn make_invoice() -> Invoice {
    Invoice {
        customer: "John Doe".to_string(),
        orders: vec![
            Order { amount: 100.00 },
            Order { amount: 250.50 },
            Order { amount: 75.25 },
        ],
        due_date: None,
    }
}

fn main() {
    println!("=== BEFORE Refactoring ===");
    print_owing_before(&mut make_invoice());

    println!("\n=== AFTER Refactoring ===");
    print_owing_after(&mut make_invoice());

    println!("\n=== Key Benefits ===");
    println!("1. Each function has a clear, single purpose");
    println!("2. Code is self-documenting through function names");
    println!("3. Individual functions can be reused and tested independently");
    println!("4. Easier to modify specific behaviors without affecting others");
}
